### PR TITLE
boards: nucleo_f103rb: enables I2C_1 and I2C_2

### LINKED ISSUE #12243: Build error with I2C driver for nucleo_f103rb: DT_ST_STM32_I2C_V1_40005400_BASE_ADDRESS' undeclared here
**Describe the bug**
Trying to compile master (65f79cdc2) with nucleo_f103rb and I2C_1

Results in 

```
In file included from /Users/mattzz/Code/zephyr/include/arch/arm/arch.h:20,
                 from /Users/mattzz/Code/zephyr/include/arch/cpu.h:15,
                 from /Users/mattzz/Code/zephyr/include/kernel_includes.h:34,
                 from /Users/mattzz/Code/zephyr/include/kernel.h:17,
                 from /Users/mattzz/Code/zephyr/include/device.h:11,
                 from /Users/mattzz/Code/zephyr/include/clock_control.h:14,
                 from /Users/mattzz/Code/zephyr/include/drivers/clock_control/stm32_clock_control.h:12,
                 from /Users/mattzz/Code/zephyr/drivers/i2c/i2c_ll_stm32.c:8:
zephyr/include/generated/generated_dts_board.h:296:45: error: 'DT_ST_STM32_I2C_V1_40005400_BASE_ADDRESS' undeclared here (not in a function); did you mean 'DT_ST_STM32_USART_40004400_BASE_ADDRESS'?
 #define DT_I2C_1_BASE_ADDRESS               DT_ST_STM32_I2C_V1_40005400_BASE_ADDRESS
                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
zephyr/include/generated/generated_dts_board.h:296:45: note: in definition of macro 'DT_I2C_1_BASE_ADDRESS'
 #define DT_I2C_1_BASE_ADDRESS               DT_ST_STM32_I2C_V1_40005400_BASE_ADDRESS
                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
zephyr/include/generated/generated_dts_board.h:303:31: error: 'DT_ST_STM32_I2C_V1_40005400_CLOCK_BITS' undeclared here (not in a function); did you mean 'DT_ST_STM32_USART_40004400_CLOCK_BITS'?
 #define DT_I2C_1_CLOCK_BITS   DT_ST_STM32_I2C_V1_40005400_CLOCK_BITS
                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
zephyr/include/generated/generated_dts_board.h:303:31: note: in definition of macro 'DT_I2C_1_CLOCK_BITS'
 #define DT_I2C_1_CLOCK_BITS   DT_ST_STM32_I2C_V1_40005400_CLOCK_BITS
                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
zephyr/include/generated/generated_dts_board.h:304:30: error: 'DT_ST_STM32_I2C_V1_40005400_CLOCK_BUS' undeclared here (not in a function); did you mean 'DT_ST_STM32_USART_40004400_CLOCK_BUS'?
 #define DT_I2C_1_CLOCK_BUS   DT_ST_STM32_I2C_V1_40005400_CLOCK_BUS
                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
zephyr/include/generated/generated_dts_board.h:304:30: note: in definition of macro 'DT_I2C_1_CLOCK_BUS'
 #define DT_I2C_1_CLOCK_BUS   DT_ST_STM32_I2C_V1_40005400_CLOCK_BUS
                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
zephyr/include/generated/generated_dts_board.h:302:45: error: 'DT_ST_STM32_I2C_V1_40005400_CLOCK_FREQUENCY' undeclared here (not in a function); did you mean 'DT_ST_STM32_USART_40004400_CLOCK_BUS'?
 #define DT_I2C_1_BITRATE                    DT_ST_STM32_I2C_V1_40005400_CLOCK_FREQUENCY
                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
zephyr/include/generated/generated_dts_board.h:302:45: note: in definition of macro 'DT_I2C_1_BITRATE'
 #define DT_I2C_1_BITRATE                    DT_ST_STM32_I2C_V1_40005400_CLOCK_FREQUENCY
                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /Users/mattzz/Code/zephyr/include/clock_control.h:14,
                 from /Users/mattzz/Code/zephyr/include/drivers/clock_control/stm32_clock_control.h:12,
                 from /Users/mattzz/Code/zephyr/drivers/i2c/i2c_ll_stm32.c:8:
zephyr/include/generated/generated_dts_board.h:299:49: error: 'DT_ST_STM32_I2C_V1_40005400_LABEL' undeclared here (not in a function); did you mean 'DT_ST_STM32_USART_40004400_LABEL'?
 #define CONFIG_I2C_1_NAME                       DT_ST_STM32_I2C_V1_40005400_LABEL
                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/mattzz/Code/zephyr/include/device.h:107:11: note: in definition of macro 'DEVICE_AND_API_INIT'
   .name = drv_name, .init = (init_fn),     \
           ^~~~~~~~
/Users/mattzz/Code/zephyr/drivers/i2c/i2c_ll_stm32.c:238:34: note: in expansion of macro 'CONFIG_I2C_1_NAME'
 DEVICE_AND_API_INIT(i2c_stm32_1, CONFIG_I2C_1_NAME, &i2c_stm32_init,
                                  ^~~~~~~~~~~~~~~~~
[35/115] Building C object zephyr/kernel/CMakeFiles/kernel.dir/work_q.c.obj
ninja: build stopped: subcommand failed.
```

**To Reproduce**
1. mkdir build; cd build
2. cmake -DBOARD=nucleo_f103rb ..
3. ninja menuconfig (see below)
4. ninja

```
$ grep I2C zephyr/.config
# CONFIG_I2C_STM32_INTERRUPT is not set
CONFIG_I2C_STM32=y
CONFIG_I2C_STM32_V1=y
CONFIG_HAS_DTS_I2C=y
CONFIG_I2C=y
# CONFIG_I2C_SLAVE is not set
# CONFIG_I2C_GPIO is not set
# CONFIG_I2C_SBCON is not set
CONFIG_I2C_INIT_PRIORITY=60
# CONFIG_I2C_0 is not set
CONFIG_I2C_1=y
# CONFIG_I2C_2 is not set
# CONFIG_I2C_3 is not set
# CONFIG_I2C_4 is not set
# CONFIG_I2C_5 is not set
# CONFIG_I2C_6 is not set
# CONFIG_I2C_7 is not set
CONFIG_USE_STM32_LL_I2C=y
```

My understanding is that I have to use I2C_1 according to `stm32f1.dtsi`

```
		i2c1: i2c@40005400 {
			compatible = "st,stm32-i2c-v1";
			clock-frequency = <I2C_BITRATE_STANDARD>;
			#address-cells = <1>;
			#size-cells = <0>;
			reg = <0x40005400 0x400>;
			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00200000>;
			interrupts = <31 0>, <32 0>;
			interrupt-names = "event", "error";
			status = "disabled";
			label= "I2C_1";
```

**Expected behavior**
Compile w/o error

**Environment (please complete the following information):**
 - OS: MacOS
 - Toolchain (e.g Zephyr SDK, ...)

```
$ cmake -GNinja -DBOARD=nucleo_f103rb ..
Zephyr version: 1.13.99
-- Found PythonInterp: /usr/local/bin/python3 (found suitable version "3.7.2", minimum required is "3.4")
-- Selected BOARD nucleo_f103rb
Parsing Kconfig tree in /Users/mattzz/Code/zephyr/Kconfig
Loading /Users/mattzz/Code/zephyr/boards/arm/nucleo_f103rb/nucleo_f103rb_defconfig as base
Configuration written to '/Users/mattzz/Code/zph_test/build/zephyr/.config'
-- Loading /Users/mattzz/Code/zephyr/boards/arm/nucleo_f103rb/nucleo_f103rb.dts as base
-- Overlaying /Users/mattzz/Code/zephyr/dts/common/common.dts
-- Cache files will be written to: /Users/mattzz/Library/Caches/zephyr
-- The C compiler identification is GNU 8.2.1
-- The CXX compiler identification is GNU 8.2.1
-- The ASM compiler identification is GNU
-- Found assembler: /usr/local/bin/arm-none-eabi-gcc
-- Performing Test toolchain_is_ok
-- Performing Test toolchain_is_ok - Success
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/mattzz/Code/zph_test/build
-- Commit SHA or Version used
```

**Additional context**
I'm not very familiar with DTS but I can't find the file that defines e.g. `DT_ST_STM32_I2C_V1_40005400_BASE_ADDRESS`. What am I missing?

### PR DESCRIPTION
**Pull request by mattzzw**
Resolves #12243

Signed-off-by: Matthias Wientapper <m.wientapper@gmx.de>

### CODE REVIEW

**Review by ydamigos — COMMENTED** [review]
Now that you enable I2C on the board, could you configure boards arduino connector (uart, spi and i2c)? Check nucleo_f746zg for reference.

**Comment by mattzzw** [comment]
@ydamigos, first I'd like to see I2C working on the nucleo board before I look at other board's configurations.

Right now, a simple i2c_read lets the system hangs in `stm32_i2c_msg_read` at
```
_STATIC_INLINE uint32_t LL_I2C_IsActiveFlag_SB(I2C_TypeDef *I2Cx)
{
  return (READ_BIT(I2Cx->SR1, I2C_SR1_SB) == (I2C_SR1_SB));
}
```
(Just learning how to debug with Eclipse...)
I2C scanner example code shows the same behavior.
But that might be related to a different matter. I will have access to a scope in couple of days again.

**Review by erwango — CHANGES_REQUESTED** [review]
Besides activating devices in dts file, you'll need to configure activation in Kconfig.defconfig (cf nucleo_f401re for instance) and configure pinmux as well.

**Comment by mattzzw** [comment]
@erwango , thanks for the help - I got I2C working finally.
I activated only I2C_1 in Kconfig.defconfig and in pinmux (as does stm32_min_dev which is best reference imho). 

Right now the default assignment is SCL/PB6, SDA/PB7.
However, in order for the board's Arduino headers to match the Arduino pinout I2C_1 needs to be remapped to SCL/PB8, SDA/PB9 via SP1_REMAP bit of the AFIO_MAPR register. 
Should PB8/PB9 be the default mapping? If yes, where would be the best spot to do that?

**Comment by erwango** [comment]
> Should PB8/PB9 be the default mapping?

This would make sense.

> If yes, where would be the best spot to do that?

Theoretically should go in `drivers/pinmux/stm32/pinmux_stm32.c`.

**Comment by mattzzw** [comment]
Hm. That would add a board dependency to `drivers/pinmux/stm32/pinmux_stm32.c` - Is that a good thing?

I succeeded remapping the pin to PB8/PB9 by adding 
`(*(volatile unsigned int *)(0x40010004)) |= 2;` (AFIO_MAPR, I2C1_REMAP bit)
 to main.c. I don't know how to bit bang properly zephyr style yet.

If PB8/PB9 are used in `pinmux.c` then `I2C1_REMAP` must be set.

I'm struggling to see 
a) where setting `2C1_REMAP`could best be located as it is clearly a specific board thing. Also I think AFIO regs can only be written when APB2 clock is enabled (RCC_APB2ENR, AFIO_EN).
b) how to use the right header files / use the correct register manipulation functions/macros. 

I know, it's only I2C but zephyr is quite new to me. :-)

**Comment by erwango** [comment]
> Hm. That would add a board dependency to `drivers/pinmux/stm32/pinmux_stm32.c` - Is that a good thing?

Indeed, we should avoid to make it as a board dependency. I first though it could be done in board pinmux.c func: pinmux_stm32_init.
But this is actually a soc dependent code. All STM32F1XX based boards that want to have this pin config will need to do this. Hence proposal of pinmux driver. But this doesn't need to be a board dependency. I think (but didn't look in detail, so I could be wrong) we could actually detect from the pinconf that this remap is required and apply it, can't we?
At least if possible, I'd like this could be tried. If too much of a burden let's set it back in pinmux_stm32_init.

> I succeeded remapping the pin to PB8/PB9 by adding
> `(*(volatile unsigned int *)(0x40010004)) |= 2;` (AFIO_MAPR, I2C1_REMAP bit)
> to main.c. I don't know how to bit bang properly zephyr style yet.
> 
> If PB8/PB9 are used in `pinmux.c` then `I2C1_REMAP` must be set.
> 
> I'm struggling to see
> a) where setting `2C1_REMAP`could best be located as it is clearly a specific board thing. Also I think AFIO regs can only be written when APB2 clock is enabled (RCC_APB2ENR, AFIO_EN).
> b) how to use the right header files / use the correct register manipulation functions/macros.
> 

There are dedicated helper function from STM32Cube SDK that might be handy here, you could check LL_GPIO_AF_EnableRemap_I2C1 for instance. You'll find it in `ext/hal/st/stm32cube/stm32f1xx/drivers/include/stm32f1xx_ll_gpio.h`

File: boards/arm/nucleo_f103rb/pinmux.c [hunk 1]
```diff
@@ -53,6 +55,11 @@ static int pinmux_stm32_init(struct device *port)
 
 	stm32_setup_pins(pinconf, ARRAY_SIZE(pinconf));
 
+	/* Enablio AFIO clock */
+	LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_AFIO);
+	/* Remap I2C1 from PB6/PB7 to PB8/PB9 for Arduino pin compatibility */
+	LL_GPIO_AF_EnableRemap_I2C1();
```
Comment (erwango) [inline boards/arm/nucleo_f103rb/pinmux.c]: Put this under #ifdef CONFIG_I2C_1

File: boards/arm/nucleo_f103rb/pinmux.c [hunk 2]
```diff
@@ -53,6 +55,11 @@ static int pinmux_stm32_init(struct device *port)
 
 	stm32_setup_pins(pinconf, ARRAY_SIZE(pinconf));
 
+	/* Enablio AFIO clock */
+	LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_AFIO);
```
Comment (erwango) [inline boards/arm/nucleo_f103rb/pinmux.c]: Is that actually required?
AFIO clock activation  done should already be done inside pinmux/gpio driver.

**Review by erwango — CHANGES_REQUESTED** [review]

File: boards/arm/nucleo_f103rb/pinmux.c (hunk 1 above)
Comment (mattzzw) [reply boards/arm/nucleo_f103rb/pinmux.c]: Good point - Thanks. I fixed that.

File: boards/arm/nucleo_f103rb/pinmux.c (hunk 2 above)
Comment (mattzzw) [reply boards/arm/nucleo_f103rb/pinmux.c]: As far as I know it is required.
I2C_1 does not work (and I2C init freezes the system) when line 59 is commented out. 
And I could not find a reference to enabling PERIPH_AFIO anywhere else in the zephyr code code.

**Comment by mattzzw** [comment]
Anything else I can do to for this PR?

File: boards/arm/nucleo_f103rb/pinmux.c [hunk 3]
```diff
@@ -49,6 +55,13 @@ static int pinmux_stm32_init(struct device *port)
 
 	stm32_setup_pins(pinconf, ARRAY_SIZE(pinconf));
 
+#ifdef CONFIG_I2C_1
+	/* Enablio AFIO clock */
```
Comment (erwango) [inline boards/arm/nucleo_f103rb/pinmux.c]: Update comment to "AFIO clock needed for AF remap"

File: boards/arm/nucleo_f103rb/nucleo_f103rb.dts [hunk 4]
```diff
@@ -74,3 +74,13 @@
 		status = "ok";
 	};
 };
+
+&i2c1 {
```
Comment (erwango) [inline boards/arm/nucleo_f103rb/nucleo_f103rb.dts]: arduino_i2c: &i2c1 {

**Review by erwango — CHANGES_REQUESTED** [review]
Last comments.
Additionally, please rebase on master

**Comment by erwango** [comment]
@mattzzw , sorry this PR has been left unreviewed for a (long) while.
Can you rebase and check CI issue?

**Pull request closed without merging**